### ISSUE TITLE
idrive.com backend

### ISSUE THREAD

**Issue by eladeyal-intel** [rclone/rclone#5229]
A backend for https://www.idrive.com would be great.

iDrive has two products - a backup space and a sync space

They have a REST API  documented here https://www.idrivesync.com/evs/web-developers-guide.htm

and a linux solution (perl mostly it seems) here https://www.idrivesync.com/evs/download.htm

iDriveSync  seems to support WebDav but I was not able to use it with rclone https://www.idrivesync.com/webdav

**Comment by ncw** [comment]
Maybe the first thing to do would be to see if we can get the idrive webdav working with rclone.

They offer 5 GB free so it should be possible to experiment without a subscription.

**Comment by eladeyal-intel** [comment]
I just cotnacted iDrive support. The Personal account does not support webDAV.

**Comment by snvoid** [comment]
It would be awesome if this can be added somehow.

**Comment by sawasy** [comment]
I found idrive 5tb drive for 1 year at 3.78 through techradar here: https://www.idrive.com/idrive/signup/el/techradar3tb?subtag=trd-us-2189611336875799300

I am a long time rclone user who has picked up this deal, so I would be happy to act as a test monkey for any development on this.

m!

**Comment by coderforlife** [comment]
They support S3 as well, see the following resources:
- https://www.idrive.com/cloud/guides/s3_developer_guide
- https://www.idrive.com/cloud/faq#2

Confusing the matter there seem to be several separate services under similar names:
- https://www.idrive.com - for backup, the free (5GB) and personal levels (5TB, $70/year) don't support WebDAV as mentioned above, higher levels may (although I cannot confirm)
- https://www.idrivesync.com/ - for syncing, there is free (5GB) and then higher levels (150GB-1TB, $35-$300/year), much more expensive, presumably since this is designed for high egress unlike the backup one. Supports [WebDAV](https://www.idrivesync.com/webdav) but is significantly limited.
- https://www.idrive.com/cloud/ - enterprise storage, 2TB is $70/year with discouraged egress, 30-day free trial, definitely supports S3

I am curious if either the idrive-backup or idrive-sync support S3 as well.

I have somewhat studied their perl scripts for idrive-backup (which may also be used for idrive-sync). Their code is a complete mess and I started making a Python replacement before finding rclone for all my other data-movement needs. Besides the perl scripts, there is a pre-compiled binary command line utility to deal with high-efficiency data transfers. The scripts only deal with the high level things like authentication, configuring, etc (all using the [REST API](https://www.idrivesync.com/evs/web-developers-guide.htm)) while the uploads and downloads are deferred to the command line program (the scripts can do upload/download with the REST API as well if hey can't find the binary).

**Comment by eladeyal-intel** [comment]
@coderforlife thanks for looking into this. To confuse things even more, the personal tier of https://www.idrive.com does include "Sync and could storage" separate than the backup storage. The perl scripts only access the "cloud backup" functionality and not the "Sync and could storage".

Tried contacting support and they they said S3 not supported for personal tier.

BTW https://www.idrive.com/cloud/ seems to be US$8 for first year

**Comment by GustavoEHennemann** [comment]
Agreed, a backend for iDrive would be really great. And they prices are really low.
I sent an email to support and I got the information that S3 is not a feature for the personal iDrive service, only for the iDrive Cloud.
So, there is no information about the communication protocol used for the personal iDrive, maybe its a specific protocol or just the same S3 hided in somewhere, maybe the experts can discover it and implement it ;-)

**Comment by lenamtl** [comment]
Hi,
I was about to use iDriveSync APi
But it does not seems to work anymore...
Can anyone tell me if this is working or not?

**Comment by dirtydevver** [comment]
a backend for this cloud provider would be amazing, the prices are good right now, but i also use rclone for all my other providers and want to keep using this rather than their dodgy scripts or windows application

**Comment by mdmmarshall67** [comment]
> I have somewhat studied their perl scripts for idrive-backup (which may also be used for idrive-sync). Their code is a complete mess and I started making a Python replacement before finding rclone for all my other data-movement needs. Besides the perl scripts, there is a pre-compiled binary command line utility to deal with high-efficiency data transfers. The scripts only deal with the high level things like authentication, configuring, etc (all using the [REST API](https://www.idrivesync.com/evs/web-developers-guide.htm)) while the uploads and downloads are deferred to the command line program (the scripts can do upload/download with the REST API as well if hey can't find the binary).

Any chance you could share the python scripts .The perl script that Idrive supply fails to work and I have had no reply for there support people

**Comment by dkrahmer** [comment]
I recently wrote an IDrive backend and API client library in C#. It is tailored for Duplicati integration but is pretty clean and concise (single file < 400 lines). It implements basic CRUD methods with upload and download having the most specialized code. There are some other methods that I did not implemented because they were not needed: move, rename, copy, file version history, etc. This code would likely be helpful as a reference to create a backend for RClone.

I was able to decipher the IDrive for Linux PERL code and figured out how to authenticate using an undocumented IDrive web API. I used the IDriveSync API documentation to implement the required client methods.

You can view my C# IDrive API code here: https://github.com/dkrahmer/duplicati/blob/DK_master/Duplicati/Library/Backend/IDrive/IDriveApiClient.cs

**Comment by bits01** [comment]
Note there is also iDrive E2 (S3 compatible) which I believe is different from the backend this issue is asking for.
https://www.idrive.com/e2/

**Comment by jamescochran** [comment]
I know this is closed but the IDrive E2 and the IDrive personal are two separate products and you even need a different login to a different TLD. It would be nice to have support for personal. I'm going to look into @dkrahmer solution and see if I can get that working though.

**Comment by milangm89** [comment]
Hi Team, It would be really great if you can implement idrive (not e2) to rclone as a backend. Please help.
https://www.idrive.com/clouddrive

**Comment by Slothoncrack** [comment]
@jamescochran Did you get a chance to look at it? Considering how cheap this service is, would be a great option to have.

**Comment by tomekit** [comment]
I might be on the wrong side, but I believe that connecting to `iDrive` not `iDrive E2` from any other app than their official is against their TOS. 

https://www.idrive.com/terms-and-conditions
> By using the Services, you agree that you:
> (d) will not access or attempt to access the Services by any means other than those provided on the IDrive® website or by an authorized third party.

Obviously if someone wants to have a go and implement it, go ahead.
